### ISSUE TITLE
no-unnecessary-type-assertion fires non-deterministically under --type-aware --type-check

### ISSUE THREAD

**Issue by Omarov3000** [oxc-project/oxc#21752]
- oxlint@1.61.0
- oxlint-tsgolint@0.22.0
- typescript@5.9.3                                                   
- bun 1.3.13, macOS 26 (Darwin 25.0.0)
                                                                      
oxlint src --type-aware --type-check --deny-warnings reports         
typescript-eslint(no-unnecessary-type-assertion)                
non-deterministically on identical input:                            
                                                                
- ~70-80% of runs report 1 warning                                   
- ~20-30% of runs report 0 warnings
- No edits between runs, single process, same args                   
                                                                      
Repro                                                                
                                                                      
https://github.com/Omarov3000/oxlint-typecheck-flaky-repro                  

git clone git@github.com:Omarov3000/oxlint-typecheck-flaky-repro.git                                                  
cd oxlint-typecheck-flaky-repro
bun install                                                          
cd libs/platform/infra/orm                                      
bun run repro

bun run repro runs the linter 50× sequentially and prints the warning-count distribution. Expected output:                      
  12 0                                                          
  38 1
                                                                      
(Exact ratio varies per machine; both outcomes are observed.)        
                                                                      
Why this matters:                                                     
                                                                
CI / pre-commit hooks configured with oxlint fail randomly on identical commits. We hit this in our monorepo. 

What we tried

- --threads=1 — still flaky (not internal worker race).              
- Reducing surface area increases flake rate (~13% in full repo, ~76%  in this minimal repro). Removing more code around the cast site     
eventually eliminates the warning entirely; the precise generic shape  that triggers it is preserved here.                                 
                                                                
The triggering code

libs/platform/infra/orm/src/schema/table.ts:84:                      

```ts  
async insert<TSelf extends this, TSelfCols extends                   
ColumnsOnly<TSelf>>(                                                 
  this: TSelf,
  data: InsertableForCols<TSelfCols>,                                
): Promise<SelectableForCols<TSelfCols>> {                           
  // The cast is load-bearing — make's TSelfCols is independently    
inferred from `this`,                                                
  // distinct from insert's TSelfCols. Removing it yields tsc TS2322 
in the original ORM.                                                 
  const dataToInsert = this.make(                               
    data as any,                                                     
  ) as SelectableForCols<TSelfCols>                                  
  return dataToInsert
}                                                                    
```       
                                                 
Related

- #20656 (CLOSED) — no-unnecessary-type-assertion false positive on  
generic functions with typeAware
- #20682 (CLOSED) — same rule, generic functions with default type   
parameters                                                           
  
The closed issues addressed content of false positives. This issue is about non-determinism of the same rule on identical input.

**Comment by fmsouza** [comment]
Confirming this reproduces in a completely unrelated codebase (React/TypeScript monorepo, not related to the original reporter's project), on the exact same `oxlint@1.61.0` pin.

Same symptom: `typescript-eslint(no-unnecessary-type-assertion)` fires non-deterministically on identical input, e.g. `document.querySelector('[role="none"]') as HTMLElement | null` and `screen.getByTestId('col') as HTMLElement` (the latter is a genuinely-redundant cast that RTL's `getByTestId<T extends HTMLElement = HTMLElement>` default already makes unnecessary — but the point is the rule doesn't fire *consistently* even on that one).

Additional data points that might help narrow this down:

- `--threads=1` does NOT fix it here either (confirming it's not an oxlint-level worker race): 25 repeated runs against a `packages/ui/src` subtree gave 18 clean / 2 with 1 warning / 5 with 2 warnings.
- Still reproduces on the latest release as of this writing, `oxlint@1.72.0`: 15 repeated runs gave 13 clean / 1 with 2 warnings / 1 with 3 warnings.
- In our case the finding is consistently confined to the same 2 files across dozens of runs (never spreads to other files with similar cast patterns elsewhere in the same subtree), which might suggest something keyed to specific type-cache entries or file processing order rather than a uniformly-random race.

Happy to share a minimal repro from our side if useful, though the affected code shapes are close to what's already posted in the linked repro repo (a cast narrowing a `querySelector`/testing-library return type, used a couple of lines before/after the assertion).

**Comment by jacopo-mistral** [comment]
Still present on current versions, and the flake rate tracks program size in a way that may be worth measuring rather than only observing.

**Versions:** `oxlint` 1.73.0 and 1.79.0, `oxlint-tsgolint` 7.0.2001, `typescript` 6.0.3, public npm only. So it survived `oxlint-tsgolint` 0.22 → 7.x.

**Repro shape** — a standalone project, 95 trivial modules plus one file holding:

```ts
declare function findByRole<T extends HTMLElement = HTMLElement>(role: string): Promise<T>;
const chooser = (await findByRole("combobox")) as HTMLSelectElement;
```

`typescript/no-unnecessary-type-assertion` fires on **9 of 40 runs** over identical bytes. Nothing varies between runs but the run. This is the shape analysed in oxc-project/tsgolint#1141 — `T` appears in no parameter, so it is inferred from the assertion.

**The count of reported violations, not just presence or absence.** This is what I think adds to the report above:

| | violations reported | runs |
|---|---|---|
| the file checked alone | 3 of 3 | 10 of 10 |
| the same file inside the 96-file program | 1 of 3 | 25 of 25 |

Same direction as "reducing surface area increases flake rate" in the original report, and it suggests the effect is not only intermittent but systematic under load: in a large program the checker reports a stable *subset*. For a CI gate that means a green run carries no information, which is worse than a flaky red one — nobody investigates a pass.

**Ruled out, each with evidence:**

| | |
|---|---|
| thread count | same rate at `--threads=1`, `2` and `4`; happens single-threaded |
| the import graph | 11/40 with the file importing nothing, 9/40 importing all 95 |
| oxlint version | 1.79.0 gives 11/40 |
| a missing tsgolint silently greening the run | it fails loudly — `Failed to find tsgolint executable`, exit 1 |
| a missing or misresolved `tsconfig.json` | the rule still reports with the tsconfig renamed away |
| the reported rule count as evidence the pass ran | `227 rules` is identical with and without `--type-aware` |

I could not narrow the trigger further than "correlates with the number of files in the program and nothing else I could vary", which is consistent with the per-node signature cache in #1141.

**Comment by lazka** [comment]
Same here, I have a public reproducer if wanted. oxlint@1.79.0, oxlint-tsgolint@7.0.2001.

**Comment by sudsy** [comment]
Another data point, from a fairly large codebase.

Setup: oxlint 1.83.0, oxlint-tsgolint 7.0.2002, 2038 files, 8 threads.

What I did:

Ran oxlint over the same unchanged tree 11 times.
Counted no-unnecessary-type-assertion findings each run: 84 84 84 86 84 86 86 86 85 84 86.
Diffed the findings run to run. Only 3 sites ever changed. The other 81 were the same in every run.
Ran it 3 more times with GOMAXPROCS=1: 84 each time, with those 3 sites silent.
The count never went below 84, only above it.

All 3 of the sites that moved are the same shape:


The signature is querySelector<E extends Element = Element>(selectors: string): E | null from lib.dom, so there are no custom generics involved.

Two of the 3 are in the same file and moved together in all 11 runs (00010111001). The third is in a different file and moved on its own (00000000100).

I'm on oxlint-tsgolint 7.0.2002, published 18 Sep, so this is the latest release as of today.
